carve/carve.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	"github.com/ipfs/go-cid"
+	carv1 "github.com/ipld/go-car"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/index"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-varint"
+	"github.com/urfave/cli/v2"
+)
+
+func main() {
+	app := &cli.App{
+		Name:  "carve",
+		Usage: "A utility for slicing CARs",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "codec",
+				Aliases: []string{"c"},
+				Usage:   "The type of index to write",
+				Value:   "CarMultihashIndexSorted",
+			},
+		},
+		Commands: []*cli.Command{
+			{
+				Name:    "index",
+				Aliases: []string{"i"},
+				Usage:   "write out the car with an index",
+				Action: func(c *cli.Context) error {
+					r, err := carv2.OpenReader(c.Args().Get(0))
+					if err != nil {
+						return err
+					}
+					defer r.Close()
+
+					var idx index.Index
+					if c.String("codec") != "none" {
+						mc, ok := multicodec.Of(c.String("codec"))
```

Comment:
probably a good time to update go-car with https://github.com/multiformats/go-multicodec/pull/51:

```
var mc multicodec.Code
if err := mc.Set(c.String("codec")); err != nil {
    return err
}
```